File: POTD/Day-17_Expression-Contains-Redundant-Bracket/Solution.cpp
class Solution {
  public:
    // =====================================================
    // ✅ OPTIMAL APPROACH (Stack-based)
    // =====================================================
    bool checkRedundancy(string &s) {

        stack<char> st;

        for (char ch : s) {
            if (ch == ')') {
                bool hasOperator = false;

                // Check contents inside brackets
                while (!st.empty() && st.top() != '(') {
                    char top = st.top();
                    st.pop();
                    if (top == '+' || top == '-' || top == '*' || top == '/')
                        hasOperator = true;
                }

                // Pop opening bracket
                if (!st.empty()) st.pop();

                // If no operator found → redundant
                if (!hasOperator)
                    return true;
            } 
            else {
                st.push(ch);
            }
        }

        return false;
    }
};

/*
========================= OTHER APPROACHES =========================

1️⃣ Brute Force
- Check every substring inside ()
- Verify presence of operator
- Time: O(n^2)
- ❌ Not efficient

---------------------------------------------------------------

2️⃣ Regex Based (Not Recommended)
- Hard to maintain
- ❌ Not interview friendly

---------------------------------------------------------------

3️⃣ Full Standalone Runnable Version

#include <iostream>
#include <stack>
using namespace std;

int main() {
    string s = "((a+b))";
    Solution sol;
    cout << sol.checkRedundancy(s);
    return 0;
}

===================================================================
*/

File: POTD/Day-17_Expression-Contains-Redundant-Bracket/Solution.py
class Solution:

    # =====================================================
    # ✅ OPTIMAL APPROACH (Stack-based)
    # =====================================================
    def checkRedundancy(self, s):

        stack = []

        for ch in s:
            if ch == ')':
                has_operator = False

                while stack and stack[-1] != '(':
                    top = stack.pop()
                    if top in '+-*/':
                        has_operator = True

                # Pop '('
                if stack:
                    stack.pop()

                if not has_operator:
                    return True
            else:
                stack.append(ch)

        return False


'''
========================= OTHER APPROACHES =========================

1️⃣ Brute Force
- Generate substrings inside ()
- Check for operators
- ❌ O(n^2)

---------------------------------------------------------------

2️⃣ Regex
- Difficult to maintain
- ❌ Not interview friendly

---------------------------------------------------------------

3️⃣ Full Standalone Runnable Version

if __name__ == "__main__":
    s = "((a+b))"
    sol = Solution()
    print(sol.checkRedundancy(s))

===================================================================
'''
